tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py

```diff
@@ -0,0 +1,47 @@
+import unittest
+
+import numpy
+
+import pytest
+
+from cupy import testing
+import cupyx.scipy.ndimage  # NOQA
+
+try:
+    import scipy.ndimage  # NOQA
+except ImportError:
+    pass
+
+
+@testing.parameterize(*testing.product({
+    'shape': [(3, 4), (2, 3, 4), (1, 2, 3, 4)],
+    'ksize': [3, 4],
+    'mode': ['reflect', 'constant', 'nearest', 'mirror', 'wrap'],
+    'cval': [0.0],
+    'origin': [-1, 0, 1],
+    'adtype': [numpy.int8, numpy.int16, numpy.int32,
+               numpy.float32, numpy.float64],
+    'wdtype': [None, numpy.int32, numpy.float64],
+    'output': [None, numpy.int32, numpy.float64],
+    'filter': ['convolve', 'correlate']
+}))
+@testing.gpu
+@testing.with_requires('scipy')
+class TestConvolveAndCorrelate(unittest.TestCase):
+
+    def _filter(self, xp, scp, a, w):
+        filter = getattr(scp.ndimage, self.filter)
+        return filter(a, w, output=self.output, mode=self.mode,
+                      cval=self.cval, origin=self.origin)
+
+    @testing.numpy_cupy_allclose(atol=1e-5, rtol=1e-5, scipy_name='scp')
+    def test_convolve_and_correlate(self, xp, scp):
+        if self.adtype == self.wdtype or self.adtype == self.output:
+            pytest.skip()
```

Comment:
Please use `raise unittest.SkipTest()` with an appropriate skip message.

Reply:
When `raise unittest.SkipTest()` was used here, it raised an exception and stopped with failure.
Any work-around?

Reply:
What error did you get? `SkipTest` is already used in some tests.

Reply:
I think that `return unittest.SkipTest()` is correct, right?

Reply:
No, `SkipTest` is an exception. https://docs.python.org/3/library/unittest.html#unittest.SkipTest

Reply:
When `return unittest.SkipTest()` is used instead as shown below,
```
@testing.numpy_cupy_allclose(atol=1e-5, rtol=1e-5, scipy_name='scp')
def test_convolve_and_correlate(self, xp, scp):
    if self.adtype == self.wdtype or self.adtype == self.output:
        return unittest.SkipTest()
```
it ends up with failure as below.
```
================================================== FAILURES ===================================================
_______________________ TestConvolveAndCorrelate_param_217.test_convolve_and_correlate ________________________
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/parameterized.py:60: in wrap
    six.raise_from(e_new.with_traceback(e.__traceback__), None)
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/parameterized.py:43: in wrap
    return method(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <cupy.testing.parameterized.TestConvolveAndCorrelate_param_217 testMethod=test_convolve_and_correlate>
args = ()
kw = {'scp': <module 'scipy' from '/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/scipy/__init__.py'>, 'xp...home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/numpy-1.17.0-py3.6-linux-x86_64.egg/numpy/__init__.py'>}
cupy_result = SkipTest(), cupy_error = None, cupy_tb = None
scipy = <module 'scipy' from '/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/scipy/__init__.py'>
numpy_result = SkipTest(), numpy_error = None, numpy_tb = None

    @functools.wraps(impl)
    def test_func(self, *args, **kw):
        if sp_name:
            kw[sp_name] = cupyx.scipy.sparse
        if scipy_name:
            kw[scipy_name] = cupyx.scipy
        kw[name] = cupy
        cupy_result, cupy_error, cupy_tb = _call_func(self, impl, args, kw)
    
        kw[name] = numpy
        if sp_name:
            import scipy.sparse
            kw[sp_name] = scipy.sparse
        if scipy_name:
            import scipy
            kw[scipy_name] = scipy
        numpy_result, numpy_error, numpy_tb = \
            _call_func(self, impl, args, kw)
    
        if cupy_error or numpy_error:
            _check_cupy_numpy_error(self, cupy_error, cupy_tb,
                                    numpy_error, numpy_tb,
                                    accept_error=accept_error)
            return
    
>       assert cupy_result.shape == numpy_result.shape
E       AssertionError: Parameterized test failed.
E       
E       Base test method: TestConvolveAndCorrelate.test_convolve_and_correlate
E       Test parameters:
E         adtype: <class 'numpy.int32'>
E         cval: 0.0
E         filter: convolve
E         ksize: 3
E         mode: reflect
E         origin: 0
E         output: None
E         shape: (3, 4)
E         wdtype: <class 'numpy.int32'>
E       
E       AttributeError: 'SkipTest' object has no attribute 'shape'

/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py:159: AssertionError
```

Reply:
Well, it looks `unittest.SkipTest()` is returned in some tests in CuPy like below.
https://github.com/cupy/cupy/blob/f2ef649cafa0553945553bc031f4d4d5aec19a55/tests/cupy_tests/test_cudnn.py#L405-L406

Reply:
Hmm, that should be incorrect. Please use `raise` instead of `return`.

Reply:
All right, but even when `raise` is used instead as below,
```
    @testing.numpy_cupy_allclose(atol=1e-5, rtol=1e-5, scipy_name='scp')
    def test_convolve_and_correlate(self, xp, scp):
        if self.adtype == self.wdtype or self.adtype == self.output:
            raise unittest.SkipTest()
```
it causes a different failure. Looks like that is AssertionError.
```
================================================== FAILURES ===================================================
_______________________ TestConvolveAndCorrelate_param_217.test_convolve_and_correlate ________________________
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/parameterized.py:60: in wrap
    six.raise_from(e_new.with_traceback(e.__traceback__), None)
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/parameterized.py:43: in wrap
    return method(*args, **kwargs)
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py:156: in test_func
    accept_error=accept_error)
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py:115: in _check_cupy_numpy_error
    self.fail(msg)
E   AssertionError: Parameterized test failed.
E   
E   Base test method: TestConvolveAndCorrelate.test_convolve_and_correlate
E   Test parameters:
E     adtype: <class 'numpy.int32'>
E     cval: 0.0
E     filter: convolve
E     ksize: 3
E     mode: reflect
E     origin: 0
E     output: None
E     shape: (3, 4)
E     wdtype: <class 'numpy.int32'>
E   
E   AssertionError: Both cupy and numpy raise exceptions
E   
E   cupy
E   Traceback (most recent call last):
E     File "/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py", line 28, in _call_func
E       result = impl(self, *args, **kw)
E     File "/home/anaruse/work/CUPY/cupy/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 46, in test_convolve_and_correlate
E       raise unittest.SkipTest()
E   unittest.case.SkipTest
E   
E   numpy
E   Traceback (most recent call last):
E     File "/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py", line 28, in _call_func
E       result = impl(self, *args, **kw)
E     File "/home/anaruse/work/CUPY/cupy/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 46, in test_convolve_and_correlate
E       raise unittest.SkipTest()
E   unittest.case.SkipTest
```
Any thought on this?

Reply:
BTW, could you tell why `pytest.skip()` should not be used here?

Reply:
There's a skipTest function call (https://docs.python.org/3.6/library/unittest.html#unittest.TestCase.skipTest) and a skipTest exception (https://docs.python.org/3.6/library/unittest.html#unittest.SkipTest). Always use the former unless you're catching the latter.

Reply:
> BTW, could you tell why `pytest.skip()` should not be used here?

I wondered too. I thought you're gradually transitioning to use pytest functionalities.

But, I can give you a legitimate reason. `pytest.skip()` is incompatible with `unittest` being the test runner (pytest-dev/pytest#5832), although in CuPy's CI services you always use pytest as runner.

Reply:
Thanks, @leofang .  But even with `skipTest`, it ends up with a failure...
```
    @testing.numpy_cupy_allclose(atol=1e-5, rtol=1e-5, scipy_name='scp')
    def test_convolve_and_correlate(self, xp, scp):
        if self.adtype == self.wdtype or self.adtype == self.output:
            return self.skipTest("skip a duplicated test")
```
```
================================================== FAILURES ===================================================
_______________________ TestConvolveAndCorrelate_param_217.test_convolve_and_correlate ________________________
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/parameterized.py:60: in wrap
    six.raise_from(e_new.with_traceback(e.__traceback__), None)
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/parameterized.py:43: in wrap
    return method(*args, **kwargs)
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py:156: in test_func
    accept_error=accept_error)
/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py:115: in _check_cupy_numpy_error
    self.fail(msg)
E   AssertionError: Parameterized test failed.
E   
E   Base test method: TestConvolveAndCorrelate.test_convolve_and_correlate
E   Test parameters:
E     adtype: <class 'numpy.int32'>
E     cval: 0.0
E     filter: convolve
E     ksize: 3
E     mode: reflect
E     origin: 0
E     output: None
E     shape: (3, 4)
E     wdtype: <class 'numpy.int32'>
E   
E   AssertionError: Both cupy and numpy raise exceptions
E   
E   cupy
E   Traceback (most recent call last):
E     File "/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py", line 28, in _call_func
E       result = impl(self, *args, **kw)
E     File "/home/anaruse/work/CUPY/cupy/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 49, in test_convolve_and_correlate
E       return self.skipTest("skip a duplicated test")
E     File "/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/unittest/case.py", line 666, in skipTest
E       raise SkipTest(reason)
E   unittest.case.SkipTest: skip a duplicated test
E   
E   numpy
E   Traceback (most recent call last):
E     File "/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/site-packages/cupy-7.0.0b3-py3.6-linux-x86_64.egg/cupy/testing/helper.py", line 28, in _call_func
E       result = impl(self, *args, **kw)
E     File "/home/anaruse/work/CUPY/cupy/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 49, in test_convolve_and_correlate
E       return self.skipTest("skip a duplicated test")
E     File "/home/anaruse/.pyenv/versions/3.6.8/lib/python3.6/unittest/case.py", line 666, in skipTest
E       raise SkipTest(reason)
E   unittest.case.SkipTest: skip a duplicated test
```
Maybe we can't use `skipTest` or `SkipTest` in the decorator `testing.numpy_cupy_allclose` ?

Reply:
https://github.com/chainer/chainer/issues/4272
I think this is the reason for not using `pytest.skip()`.
I don't come up with a reason for using `SkipTest` instead of `skipTest()` though.

Reply:
I just realized `unittest.TestCase.skipTest()` is an instance method that requires `self`.
In some (relatively rare) cases this is very inconvenient.
In contrary `raise SkipTest('...')` is always correct. That's why we suggest using `SkipTest`.

Reply:
How about just calling `self.skipTest()`?
https://github.com/cupy/cupy/blob/db9eec54925bd9d1cf3732e70c91d0295e4ca039/tests/cupy_tests/testing_tests/test_parameterized.py#L9
> Maybe we can't use `skipTest` or `SkipTest` in the decorator `testing.numpy_cupy_allclose` ?

Yes, I think `return skipTest()` wouldn't work, because `numpy_cupy_allclose` requires an array to be returned.

Reply:
> [chainer/chainer#4272](https://github.com/chainer/chainer/issues/4272)
> I think this is the reason for not using `pytest.skip()`.
> I don't come up with a reason for using `SkipTest` instead of `skipTest()` though.

This should be documented in the contribution guide...Some of my recent PRs also used `pytest.skip()`...

Reply:
I mean, in some cases we have to pass around `self` just to skip a test. If the skip is caused after layers of function calls, it's cumbersome to do that.

That's similar to one of the reasons why I hate `self.assertEqual` in `unittest` and love `assert`s in pytest.

Reply:
> I don't come up with a reason for using `SkipTest` instead of `skipTest()` though.

I think the latter is recommended in the Python documentation:
> Usually you can use TestCase.skipTest() or one of the skipping decorators instead of raising this directly.

My interpretation of the documentation is that only `setUp()` and `tearDown()` can raise `SkipTest`, not in the test method (again, unless you catch it).

Reply:
I'm still wondering what to do for this?
One solution is to give up to skip redundant tests at expense of longer testing time. Would it be OK for you?

Reply:
How about `return xp.array(True)`? This trick is used in other tests (like `test_ndarray_elementwise_op.py`).

Reply:
Btw I made an issue for supporting test skips in `numpy_cupy_allclose`: https://github.com/cupy/cupy/issues/2492

Reply:
Thanks for your suggestion, @niboshi.  I modified the PR so that it uses your work-around (`return xp.array(True)`) to skip tests.